applier/src/main/java/com/google/archivepatcher/applier/zip/CustomDeflater.java

```diff
@@ -0,0 +1,530 @@
+                int n = deflateBytes(zsRef.address(), b, off, len, flush);
+                bytesWritten += n;
+                bytesRead += (thisLen - this.len);
+                return n;
+            }
+            throw new IllegalArgumentException();
+        }
+    }
+
+    /**
+     * Returns the ADLER-32 value of the uncompressed data.
+     *
+     * @return the ADLER-32 value of the uncompressed data
+     */
+    public int getAdler() {
+        synchronized (zsRef) {
+            ensureOpen();
+            return getAdler(zsRef.address());
+        }
+    }
+
+    /**
+     * Returns the total number of uncompressed bytes input so far.
+     *
+     * <p>Since the number of bytes may be greater than
+     * Integer.MAX_VALUE, the {@link #getBytesRead()} method is now
+     * the preferred means of obtaining this information.</p>
+     *
+     * @return the total number of uncompressed bytes input so far
+     */
+    public int getTotalIn() {
+        return (int) getBytesRead();
+    }
+
+    /**
+     * Returns the total number of uncompressed bytes input so far.
+     *
+     * @return the total (non-negative) number of uncompressed bytes input so far
+     * @since 1.5
+     */
+    public long getBytesRead() {
+        synchronized (zsRef) {
+            ensureOpen();
+            return bytesRead;
+        }
+    }
+
+    /**
+     * Returns the total number of compressed bytes output so far.
+     *
+     * <p>Since the number of bytes may be greater than
+     * Integer.MAX_VALUE, the {@link #getBytesWritten()} method is now
+     * the preferred means of obtaining this information.</p>
+     *
+     * @return the total number of compressed bytes output so far
+     */
+    public int getTotalOut() {
+        return (int) getBytesWritten();
+    }
+
+    /**
+     * Returns the total number of compressed bytes output so far.
+     *
+     * @return the total (non-negative) number of compressed bytes output so far
+     * @since 1.5
+     */
+    public long getBytesWritten() {
+        synchronized (zsRef) {
+            ensureOpen();
+            return bytesWritten;
+        }
+    }
+
+    /**
+     * Resets deflater so that a new set of input data can be processed.
+     * Keeps current compression level and strategy settings.
+     */
+    public void reset() {
+        synchronized (zsRef) {
+            ensureOpen();
+            reset(zsRef.address());
+            finish = false;
+            finished = false;
+            off = len = 0;
+            bytesRead = bytesWritten = 0;
+        }
+    }
+
+    /**
+     * Closes the compressor and discards any unprocessed input.
+     * This method should be called when the compressor is no longer
+     * being used, but will also be called automatically by the
+     * finalize() method. Once this method is called, the behavior
+     * of the Deflater object is undefined.
+     */
+    public void end() {
+        synchronized (zsRef) {
```

Comment:
The problem is that `CustomerDeflater` inherits from `Deflater` without actually making use of any of its fields, though those fields do get initialised. Even without memory leaks, is less efficient than it could be.

Inspired by your solution, we made our own, where we introduced an interface `IDeflater`, to avoid this issue with inheriting from `Deflater`. We'll publish our solution properly in the next couple of days, but you can already view [its sources](https://github.com/EIDU/archive-patcher-android/blob/main/lib/src/main/java/com/eidu/zip/AndroidDeflater.java) if you're interested.

Reply:
The problem is that `CustomerDeflater` inherits from `Deflater` without actually making use of any of its fields, though those fields do get initialised. Even without memory leaks, is less efficient than it could be.

Inspired by your solution, we made our own, where we introduced an interface `IDeflater`, to avoid this issue with inheriting from `Deflater`. We'll publish our solution properly in the next couple of days, but you can already view [its sources](https://github.com/EIDU/archive-patcher-android/blob/main/lib/src/main/java/com/eidu/zip/AndroidDeflater.java) if you're interested.